src/index.ts

```diff
@@ -22,37 +52,73 @@ export const goSync = <T, E extends Error>(fn: () => T): GoResult<T, E> => {
+  }
+  return retryFn(() => fn, attemptOptions);
+};
+
+const retryFn = async <T, E extends Error>(
+  fn: () => Promise<T>,
+  attemptOptions: AttemptOptions<T>
+): Promise<GoResult<T, E>> => {
+  return retry(fn, attemptOptions)
+    .then(success)
+    .catch((err) => {
+      return createGoError(err);
+    });
 };
 
-export const go = async <T, E extends Error>(fn: Promise<T> | (() => Promise<T>)): Promise<GoResult<T, E>> => {
+export const go = async <T, E extends Error>(
+  fn: Promise<T> | (() => Promise<T>),
+  options?: PromiseOptions
+): Promise<GoResult<T, E>> => {
+  const attemptOptions: AttemptOptions<any> = {
+    delay: options?.retryDelayMs || 200,
+    maxAttempts: (options?.retries || 0) + 1,
+    initialDelay: 0,
+    minDelay: 0,
+    maxDelay: 0,
+    factor: 0,
+    timeout: options?.timeoutMs || 0,
+    jitter: false,
+    handleError: null,
+    handleTimeout: options?.timeoutMs
+      ? async (context: AttemptContext, options: AttemptOptions<T>) => {
+          if (context.attemptsRemaining > 0) {
+            const res = await retryFnWrapper(fn, { ...options, maxAttempts: context.attemptsRemaining });
+
+            if (res.success) {
+              return res.data;
+            } else {
+              throw res.error;
+            }
+          }
+          throw new Error(`Operation timed out after final retry`);
+        }
+      : null,
+    beforeAttempt: null,
+    calculateDelay: null,
+  };
+
   // We need try/catch because `fn` might throw sync errors as well
   try {
-    if (typeof fn === 'function') {
-      return fn()
-        .then(success)
-        .catch((err) => createGoError(err));
-    }
-    return fn.then(success).catch((err) => createGoError(err));
+    return retryFnWrapper(fn, attemptOptions);
   } catch (err) {
     return createGoError(err);
   }
 };
 
-// NOTE: This needs to be written using 'function' syntax (cannot be arrow function)
-// See: https://github.com/microsoft/TypeScript/issues/34523#issuecomment-542978853
-export function assertGoSuccess<T>(result: GoResult<T>): asserts result is GoResultSuccess<T> {
-  if (!result.success) {
-    throw result.error;
-  }
-}
+export const retryGo = <T>(fn: Promise<T> | (() => Promise<T>), options?: PromiseOptions) =>
+  go(fn, { retries: DEFAULT_RETRIES, ...options });
 
-// NOTE: This needs to be written using 'function' syntax (cannot be arrow function)
-// See: https://github.com/microsoft/TypeScript/issues/34523#issuecomment-542978853
-export function assertGoError<E extends Error>(result: GoResult<any, E>): asserts result is GoResultError<E> {
-  if (result.success) {
-    throw new Error('Assertion failed. Expected error, but no error was thrown');
-  }
-}
+export const timeoutGo = <T>(fn: Promise<T> | (() => Promise<T>), options?: PromiseOptions) =>
+  go(fn, { timeoutMs: DEFAULT_RETRY_TIMEOUT_MS, ...options });
+
+export const retryTimeoutGo = <T>(fn: Promise<T> | (() => Promise<T>), options?: PromiseOptions) => {
```

Comment:
Why there is this function with hardcoded retries?

Reply:
Actually do we need all of these functions? Isn't just optional argument in the go and goSync enough? cc: @amarthadan ?